### DISCUSSION TITLE
Ultimate NextJS Fast Refresh and @xstate/react setup

### DISCUSSION THREAD

**Discussion by milanbgd011** [statelyai/xstate#2023] [Show and tell]
I was frustrated that I could not make this work, I did find this custom useMachine hook that did the job for simple machine that did not include "spawned" machines: https://github.com/davidkpiano/xstate/issues/995#issuecomment-636351806

When developing app in NextJS using Xstate, NextJS should reload machine from start if I am changing the machine, that is correct. BUT, when you finish your machine and then try to create/stylize frontend using that machine, it means a lot of tiny CSS changes while machine is in specific state with specific values in context. 

So we want:
1. if we change the machine >> RELOAD MACHINE FULL
2. if we change HTML/CSS that renders using machine's state and context >> DO NOT RELOAD MACHINE, FAST REFRESH HTML/CSS

### Fast Refresh using "useMachine" scenario?

How to achieve this for "Menu" component?
1. create folder "Menu" and go into that folder
2. create "_machine.ts" and place your machine there
3. create "Menu.tsx" which uses only props and holds all HTML/CSS stuff, which exports `<Menu />` component
4. create "MenuUseMachine.tsx" and just initialize your machine there and return `<Menu context={context} send={send} />`

We almost never change MenuUseMachine.tsx because it just initializes useMachine and passes it to the Menu.tsx, but if we do anyhow we will get full machine reload. If we change _machine.ts when configuring our machine, we will get full machine reload. BUT if we change "Menu.tsx" we will get Fast Refresh for HTML/CSS only!!! Which means we can stylize HTML/CSS while preserving current state and context.

This happens because NextJS injects changes "per file", so if we change Menu.tsx, there are no "useMachine" or machine itself stuff there, just HTML/CSS, and the machine lives one level up and is not replaced, which means we get the same machine state and context and we can use all the benefits of Fast Refresh with Xstate without any custom hooks. 

I am spending ~90% in HTML and CSS and I really need my machine to hold state and context intact while I tune these.

### Fast Refresh using "useActor" scenario while reusing "Menu.tsx" where HTML/CSS lives?

I ran into second specific case, I needed to reuse the same component with the same machine but as spawned machine within some other parent machine, and I needed the same "Fast Refresh" when doing HTML/CSS, so this is achileved like this:

1. in the "Menu" folder mentioned above create "MenuUseActor.tsx" where "useActor" is configured and return `<Menu context={context} send={send} />`

### What we get from this setup?
1. REUSE Menu.tsx in useMachine and useActor scenarios
2. FULL MACHINE RELOAD: if machine/useMachine/useActor files are changed
3. PRESERVE STATE/CONTEXT, FAST REFRESH HTML/CSS: if only HTML/CSS is changed in "Menu.tsx" component

Would like to see opinions from others about this matter, this is as far as my knowledge goes, but I am very happy that I achieved "holy grail" of frontend:  NextJS Fast Refresh and Xstate to work without any additional external code and to work perfectly.

This might be obvious for advanced users, but I could not find single piece of information about this anywhere and was struggling to make this work. I hope someone will run into this when dealing with this problem.

### Example

**_machine.ts** ---> no Fast Refresh when altering this file, just refresh entire page
```js
import { Machine } from 'xstate';
export const machineForm = Machine({ 
  id: 'machineAdmin',
  initial: 'ssr',
  states: {
    // ssr is inital state which renders the same on server and on client if needed
    // but only client will trigger "START" event and make machine alive
    // this prevents hydration problems of mismatched HTML
    ssr: {
      on: {
        START: 'idle',
      },
    },
    idle: {}
  }
})
```

**FormUseMachine.ts** ---> this file almost never changes, not important for Fast Refresh
```js
import { machineForm } from './_machine';
import { useMachine } from '@xstate/react';

export default function FormUseMachine() {
  const machine = useMachine(machineForm);   // initialize useMachine hook
  const [{ context, matches }, send]: any = machine || [{}];   // destructure and rename params from machine to be passed to the React component 

  useEffect(() => {
    send({ type: 'START' );
  }, []);

  return (
    <>
      <Form context={context} matches={matches} send={send} />
    </>
  );
}
```

**Form.ts** ---> any change in this file will work with NextJS Fast Refresh and will keep state/context during changes
```javascript
export default function Form(props) {
  const { context, matches, send } = props || {};
  return (
    <>
      ....
    </>
  );
}
```

**Comment by pke** [comment]
Not sure I can follow you here. Would you be so kind to post some code snippets too?
Fast refresh is broken in nextjs apps.

**Reply by milanbgd011** [reply]
I added example to the top comment! Hope it helps.

**Reply by milanbgd011** [reply]
I updated all once again, now it looks clean. Have fun!
